Pathfinding behavior update

Cwhizard opened the pull request:
I am currently using the steam name The Great White King, so this is my code, which I posted to the forum last night.  I have additionally made the change to the salvaging code to use the same algorithm.  Since I basically see salvaging as another form of mining.

The changes add no new features, they merely improve the pathfinding quite a lot.  I haven't made actual measurements, but anecdotally the mining at least seems to be about 4 times faster due to avoiding the zig zag condition.

There is one issue, in that sometimes the script will not target the nearest asteroid to the ship when starting.  This has only happened once, and the asteroid was less than 1Km from the ship spawn point and was targeted properly down the line.  I have no earthly idea why this happened and it has not repeated since.  I suspect this would have happened with the old code as well since on script start, it basically uses the old method to target the first asteroid.

Cwhizard committed:
Update mineCommand.lua

Cwhizard committed:
Modified targeting behavior to increase optimization of salvaging and mining routes (Nearest Neighbor)

It's not the most optimal solution but it is significantly more effective than the "closest to ship" solution.  Technically this implements the Nearest Neighbor algorithm as applied to the travelling salesman problem.  Which is fancy talk for it selects the asteroid or wreck closest to the one it just finished, rather than zig zagging back and forth across the sector wasting time instead of mining/salvaging.

Actual in gae increase in harvesting rates have not been calculated, but I went from taking over an hour to mine a system to about 15 minutes.

Laserzwei commented:
I appreciate your efforts in improving the carrier command mod. 
There is however no "pathing bug", as you state it on the Avorion forum post. I chose to let mining-/salvaging-fighters go after the closest target to the mothership for 2 reasons:

1. fighters close to the mothership can be docked fast, easily and safe in case of any emerging danger
2. When mining or salvaging with fighters in your playership, you most likely want to get the valuable stuff first (maybe not even the iron/titanium etc. at all !). With the current pathing you just place your ship near the wanted target and the fighters will go after it. With clever planing of your own route the fighters will waste no time on unwanted routes or targets. This however requires planing ahead from the player. I can see a big benefit in the Nearest-Neighbor code for NPC/out-of-sector mining/salvaging ships.

Since the current implementation has its purpose, I refuse to overwrite it with your changes as they are now.
Please add a config option to optionally enable the NN-pathfinding via a checkbox. You can see examples of checkboxes here : https://github.com/Laserzwei/CarrierCommander/blob/59b6370a482a8ea91e611928f228c93052784d35/mods/CarrierCommander/scripts/entity/ai/mineCommand.lua#L38-L43 and here: https://github.com/Laserzwei/CarrierCommander/blob/59b6370a482a8ea91e611928f228c93052784d35/mods/CarrierCommander/scripts/entity/ai/aggressiveCommand.lua#L64-L76
Alternative you could create a very new command plugin as well.

Cwhizard committed:
Update aggressiveCommand.lua

Added checkbox for aggressive targeting, which lets the player select between targeting the ship closest to the players ship, or the ship closest to the squad.  Currently does nothing as I need to read through the targeting code for combat fighters more and test it in game before commit.

Cwhizard committed:
Merge pull request #1 from Cwhizard/Cwhizard-patch-1

Update aggressiveCommand.lua

Cwhizard committed:
Update CarrierCommanderConfig.lua

Because my changes are becoming non-trivial in terms of time and effort and usefulness to the project

Laserzwei commented:
I shouldn't accept this pull request, because you literally ignored my request and help to keep the current pathfinding. But oh well I'm doing it myself then
